### PR TITLE
Python bindings (win, linux, mac)

### PR DESCRIPTION
**Pull request by Stayer**
### Description of the Change

This PR allows make python wheels distributable packages on Windows, Linux and Mac OS from automaticaly generated artifacts

### Benefits

Simple and faster distribution of python packages

### Possible Drawbacks 

* Python version - 3.5.5 and 2.7.15
* Only for x86_64 systems

- [x] Versioning (for wheels) is needed
- [x] Documentation for PyPI package
- [x] Different repositories for release/dev/nightly builds
- [x] Change SWIG generation for .py files
- [x] Upload only Release wheels to artifact server
- [x] Test import module on "clean" OS  

### About Nexus
https://nexus.soramitsu.co.jp/ - our new PyPI, Maven, artifacts server

https://nexus.soramitsu.co.jp/repository/pypi-develop/ - developer wheels
https://nexus.soramitsu.co.jp/repository/pypi-release/ - release wheels
https://nexus.soramitsu.co.jp/repository/pypi-develop-nightly/ - develop nightly wheels

#### Example
```
pip install -i https://nexus.soramitsu.co.jp/repository/pypi-{develop,release,develop-nightly}/ iroha-hl
```

#### Add to ~/.pypirc
```
[distutils]
index-servers=
    pypi-develop
    pypi-release
    pypi-develop-nightly
   
[pypi-develop]
repository: https://nexus.soramitsu.co.jp/repository/pypi-develop/
[pypi-release]
repository: https://nexus.soramitsu.co.jp/repository/pypi-release/
[pypi-dev-nightly]
repository: https://nexus.soramitsu.co.jp/repository/pypi-develop-nightly/

pip install -r pypi-develop iroha-hl
```
Reference PR: #1454

### CODE REVIEW

**Comment by bakhtin** [comment]
- Please confirm whether generated bindings works on Python 3.5.x.
- In order to differentiate between different build types, I suggest to make changes in Jenkinsfile global variables section and add a boolean variable called 'Nightly'. Release and dev build types should correspond to 'master' and 'develop' branch respectively.
- Ping @luckychess to help with SWIG patching so resulting bindings use proper imports

**Comment by Stayer** [comment]
python 3 version was changed to 3.5.x

**Comment by Stayer** [comment]
wheel name:
${env.GIT_TAG_NAME}-${env.GIT_BRANCH}-${env.BUILD_NUMBER}

**Comment by Stayer** [comment]
Tested on:
OS: Windows 7 X86_64
Python 3.5.5

Log:
```
Python 3.6.5 :: Anaconda, Inc.

(base) C:\Users\tester\Downloads>activate py3.5

(py3.5) C:\Users\tester\Downloads>pip install iroha-develop-cp35-cp35m-win_amd
64.whl
Processing c:\users\tester\downloads\iroha-develop-cp35-cp35m-win_amd64.whl
Requirement already satisfied: grpcio>=1.12.1 in c:\users\tester\anaconda3\env
s\py3.5\lib\site-packages (from iroha==develop) (1.12.1)
Requirement already satisfied: protobuf>=3.5.2.post1 in c:\users\tester\anacon
da3\envs\py3.5\lib\site-packages (from iroha==develop) (3.6.0)
Requirement already satisfied: six>=1.5.2 in c:\users\tester\anaconda3\envs\py
3.5\lib\site-packages (from grpcio>=1.12.1->iroha==develop) (1.11.0)
Requirement already satisfied: setuptools in c:\users\tester\anaconda3\envs\py
3.5\lib\site-packages (from protobuf>=3.5.2.post1->iroha==develop) (39.2.0)
Installing collected packages: iroha
  Found existing installation: iroha null-nightly
    Uninstalling iroha-null-nightly:
      Successfully uninstalled iroha-null-nightly
Successfully installed iroha-develop

(py3.5) C:\Users\tester\Downloads>python
Python 3.5.5 |Anaconda, Inc.| (default, Apr  7 2018, 04:52:34) [MSC v.1900 64 bi
t (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import iroha
>>> dir(iroha)
['AddAssetQuantity', 'AddPeer', 'AddSignatory', 'AppendRole', 'Blob', 'Blob_from
HexString', 'Blob_swigregister', 'Block', 'BlocksQuery', 'BlocksQuery_swigregist
er', 'ByteVector', 'ByteVector_swigregister', 'COMMITTED', 'Command', 'CreateAcc
ount', 'CreateAsset', 'CreateDomain', 'CreateRole', 'DESCRIPTOR', 'DetachRole',
'GetAccount', 'GetAccountAssetTransactions', 'GetAccountAssets', 'GetAccountDeta
il', 'GetAccountTransactions', 'GetAssetInfo', 'GetRolePermissions', 'GetRoles',
 'GetSignatories', 'GetTransactions', 'GrantPermission', 'GrantablePermissionSet
', 'GrantablePermissionSet_swigregister', 'Grantable_COUNT', 'Grantable_kAddMySi
gnatory', 'Grantable_kRemoveMySignatory', 'Grantable_kSetMyAccountDetail', 'Gran
table_kSetMyQuorum', 'Grantable_kTransferMyAssets', 'Hash', 'HashVector', 'HashV
ector_swigregister', 'Hash_fromHexString', 'Hash_swigregister', 'Header', 'IntVe
ctor', 'IntVector_swigregister', 'Keypair', 'Keypair_swigregister', 'MST_EXPIRED
', 'ModelBlocksQueryBuilder', 'ModelBlocksQueryBuilder_swigregister', 'ModelCryp
to', 'ModelCrypto_swigregister', 'ModelProtoBlocksQuery', 'ModelProtoBlocksQuery
_swigregister', 'ModelProtoQuery', 'ModelProtoQuery_swigregister', 'ModelProtoTr
ansaction', 'ModelProtoTransaction_swigregister', 'ModelQueryBuilder', 'ModelQue
ryBuilder_swigregister', 'ModelTransactionBuilder', 'ModelTransactionBuilder_swi
gregister', 'NOT_RECEIVED', 'PrivateKey', 'PrivateKey_swigregister', 'PublicKey'
, 'PublicKey_swigregister', 'Query', 'QueryPayloadMeta', 'Query_swigregister', '
RemoveSignatory', 'RevokePermission', 'RolePermissionSet', 'RolePermissionSet_sw
igregister', 'Role_COUNT', 'Role_kAddAssetQty', 'Role_kAddMySignatory', 'Role_kA
ddPeer', 'Role_kAddSignatory', 'Role_kAppendRole', 'Role_kCreateAccount', 'Role_
kCreateAsset', 'Role_kCreateDomain', 'Role_kCreateRole', 'Role_kDetachRole', 'Ro
le_kGetAllAccAst', 'Role_kGetAllAccAstTxs', 'Role_kGetAllAccDetail', 'Role_kGetA
llAccTxs', 'Role_kGetAllAccounts', 'Role_kGetAllSignatories', 'Role_kGetAllTxs',
 'Role_kGetBlocks', 'Role_kGetDomainAccAst', 'Role_kGetDomainAccAstTxs', 'Role_k
GetDomainAccDetail', 'Role_kGetDomainAccTxs', 'Role_kGetDomainAccounts', 'Role_k
GetDomainSignatories', 'Role_kGetMyAccAst', 'Role_kGetMyAccAstTxs', 'Role_kGetMy
AccDetail', 'Role_kGetMyAccTxs', 'Role_kGetMyAccount', 'Role_kGetMySignatories',
 'Role_kGetMyTxs', 'Role_kGetRoles', 'Role_kReadAssets', 'Role_kReceive', 'Role_
kRemoveMySignatory', 'Role_kRemoveSignatory', 'Role_kSetDetail', 'Role_kSetMyAcc
ountDetail', 'Role_kSetMyQuorum', 'Role_kSetQuorum', 'Role_kSubtractAssetQty', '
Role_kTransfer', 'Role_kTransferMyAssets', 'STATEFUL_VALIDATION_FAILED', 'STATEF
UL_VALIDATION_SUCCESS', 'STATELESS_VALIDATION_FAILED', 'STATELESS_VALIDATION_SUC
CESS', 'SetAccountDetail', 'SetAccountQuorum', 'Signature', 'SignatureVector', '
SignatureVector_swigregister', 'Signature_swigregister', 'Signed', 'Signed_swigr
egister', 'StringVector', 'StringVector_swigregister', 'SubtractAssetQuantity',
'SwigPyIterator', 'SwigPyIterator_swigregister', 'ToriiResponse', 'Transaction',
 'Transaction_swigregister', 'TransferAsset', 'TxStatus', 'TxStatusRequest', 'Un
signedBlockQuery', 'UnsignedBlockQuery_swigregister', 'UnsignedQuery', 'Unsigned
Query_swigregister', 'UnsignedTx', 'UnsignedTx_swigregister', '__builtins__', '_
_cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__p
ath__', '__spec__', '_iroha', 'block__pb2', 'block_pb2', 'commands__pb2', 'comma
nds_pb2', 'descriptor_pb2', 'endpoint_pb2', 'enum_type_wrapper', 'google_dot_pro
tobuf_dot_empty__pb2', 'iroha', 'isValid', 'permissionFor', 'primitive__pb2', 'p
rimitive_pb2', 'queries__pb2', 'queries_pb2', 'responses__pb2', 'responses_pb2',
 'sys', 'toBinaryString']
>>>
```

File: docs/source/api/python_library.rst [hunk 1]
```diff
@@ -1,6 +1,76 @@
...
+
+
+Pulling from PyPI
+---------------
+
+You can download our PyPI package from our Nexus server (instead of pypi.org). To do this, please specify ``-i`` parameter:
+
+.. code-block:: shell
+
+    pip install -i https://nexus.soramitsu.co.jp/repository/pypi-{develop,release,develop-nightly}/ iroha
```
Comment (bakhtin) [inline docs/source/api/python_library.rst]: Please make it directly copyable without any modifications. E.g., `pip install -i https://nexus.soramitsu.co.jp/repository/pypi-develop/ iroha`. I suggest to split it into three lines with annotations on what each repository consists of (i.e., `developer` build from `develop` branch, etc.)

File: docs/source/api/python_library.rst [hunk 2]
```diff
@@ -1,6 +1,76 @@
...
+
+This method allows pushing to repo (only for contributors). Twine package is required.
+
+.. code-block:: shell
+
+    twine upload --repository-url https://nexus.soramitsu.co.jp/repository/pypi-{develop,release,develop-nightly}/ /path/to/iroha/wheel/iroha.whl
+
+Adding iroha PyPI repositories
+---------------
+You can simple add our Nexus server of Python packages distribution.
```
Comment (bakhtin) [inline docs/source/api/python_library.rst]: simply

File: .jenkinsci/bindings.groovy [hunk 3]
```diff
@@ -130,4 +131,49 @@ def doAndroidBindings(abiVersion) {
   return artifactsPath
 }
 
+def doPythonWheels(os, buildType) {
+  def version;
```
[the remark this answers is not in the document]
Comment (Stayer) [reply .jenkinsci/bindings.groovy]: Can't define it (windows build)
https://jenkins.soramitsu.co.jp/blue/organizations/jenkins/iroha%2Firoha-hyperledger/detail/PR-1457/168/pipeline/

File: Jenkinsfile [hunk 4]
```diff
@@ -460,6 +460,44 @@ pipeline {
             }
           }
         }
+        stage('MacOS bindings') {
+          when {
+            beforeAgent true
+            expression { return params.x86_64_macos }
+          }
+          agent { label 'mac-173' }
```
Comment (bakhtin) [inline Jenkinsfile]: should work on any mac agent we have (any `mac` label)

File: docs/source/api/python_library.rst [hunk 5]
```diff
@@ -1,6 +1,105 @@
...
-This section only contains code reference (autogenerated docs are possible with a description of arguments and return values)
 
-.. Attention:: Contents are missing for now.
\ No newline at end of file
+About Nexus
+---------------
+https://nexus.soramitsu.co.jp/ - our new PyPI, Maven, artifacts server with precompilled librares.
+You can download binaries without compilation on the client side. It helps to save you time and disk space for installing dependencies like libboost, libz, etc.
+
+Now our packages are distributed only for ``Windows x86_64`` and ``Linux x86_64`` with ``Python 2.7.x`` and ``Python 3.5.x``. In future it will be available on ``Mac OS``.
```
Comment (bakhtin) [inline docs/source/api/python_library.rst]: Already available for MacOS?

File: docs/source/api/python_library.rst [hunk 6]
```diff
@@ -1,6 +1,105 @@
...
+Pulling from PyPI
+---------------
+
+You can download our PyPI package from our Nexus server (instead of pypi.org). To do this, please specify ``-i`` parameter:
+
+From develop
+
+.. code-block:: shell
+
+    pip install -i https://nexus.soramitsu.co.jp/repository/pypi-develop/ iroha
```
Comment (bakhtin) [inline docs/source/api/python_library.rst]: Does not work on a clean Ubuntu 18.04 installation (neither python3 nor python2). Please verify

**Pull request closed without merging**